### PR TITLE
Reset password with validation code

### PR DESCRIPTION
**Pull request by emmaielle**
#### :page_facing_up: Description:

Reset password feature.

---

#### :pushpin: Notes:

- For testing, use the development endpoint for MiCambio, since it follows the same steps.
- Styling should not be final

---

#### :heavy_check_mark: Tasks:
 
Includes three stages:

- Email input for sending verification code
- Validation of said code
- Input of new password

---

#### :play_or_pause_button: Demo:
https://www.loom.com/share/4b1e1dd0d75642b6a14afe031b9f231c

@loopstudio/react-devs

### CODE REVIEW

**Comment by mconiglio** [comment]
@emmaielle https://github.com/loopstudio/rails-api-boilerplate/pull/165 this PR got merged

**Comment by donatoaguirre24** [comment]
Please update the demo video

File: src/helpers/errors.js [hunk 1]
```diff
@@ -7,10 +7,14 @@ export function handleErrors({ attributesErrors, errors }, setError) {
...
     setError(
-      Object.keys(attributesErrors).map((fieldName) => ({
-        message: capitalize(attributesErrors[fieldName][0]),
-        name: fieldName,
-      }))
+      Object.keys(attributesErrors).map((fieldName) => {
+        return {
+          message: attributesErrors?.[fieldName].length
+            ? capitalize(attributesErrors?.[fieldName][0])
+            : '',
```
[the remark this answers is not in the document]
Comment (mconiglio) [reply src/helpers/errors.js:14]: image
this is caused because the api is sending the empty array for the `reset_password_token` "attribute" when it doesn't have errors, instead of not sending the key at all

File: src/components/ForgotPassword/EmailForm.js [hunk 2]
```diff
@@ -2,52 +2,61 @@ import React, { useState } from 'react';
...
   const validationSchema = object().shape({
     email: string()
       .required(intl.messages['common.required'])
       .email(intl.messages['common.invalidEmail']),
   });
 
-  const formMethods = useForm({ validationSchema });
+  const formMethods = useForm({
+    validationSchema,
+  });
```
Comment (mwolman) [inline src/components/ForgotPassword/EmailForm.js:27]: Why this change? 🤔

File: src/components/ForgotPassword/TokenForm.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import { useForm } from 'react-hook-form';
+import { useIntl, FormattedMessage } from 'react-intl';
+import { object, string } from 'yup';
+import PropTypes from 'prop-types';
+
+import AuthService from 'services/AuthService';
+import Form from 'components/Form';
+import Loading from 'components/Loading';
+import { handleErrors } from 'helpers/errors';
+import { RESET_PASSWORD_STEPS } from './ForgotPassword';
```
Comment (mwolman) [inline src/components/ForgotPassword/TokenForm.js:11]:
```suggestion
import Form from 'components/Form';
import Loading from 'components/Loading';
import { handleErrors } from 'helpers/errors';
import AuthService from 'services/AuthService';
import { RESET_PASSWORD_STEPS } from './ForgotPassword';
```

File: src/locales/messages/es-ES.js [hunk 4]
```diff
@@ -3,34 +3,45 @@ export default {
...
     errorMessage: 'Ha ocurrido un error',
     firstName: 'Nombre',
     forgotPassword: '¿Olvidaste tu contraseña?',
+    forgotPasswordChange: 'Paso 3: Ingresa tu nueva contraseña',
+    forgotPasswordEmailResend: '¿No recibiste un código?',
+    forgotPasswordEmailSent:
+      'Paso 2: Ingresa el código de verificacion que hemos enviado a tu correo',
     forgotPasswordLegend:
-      'Escribe tu email y te ayudaremos a recuperar tu contraseña',
+      'Paso 1: Escribe tu email y te ayudaremos a recuperar tu contraseña',
```
Comment (mwolman) [inline src/locales/messages/es-ES.js:20]:
```suggestion
      'Paso 1: Ingresa tu correo y te ayudaremos a recuperar tu contraseña',
```

File: src/locales/messages/es-ES.js [hunk 5]
```diff
@@ -3,34 +3,45 @@ export default {
...
     goBackHome: 'Volver al inicio',
     invalidEmail: 'Email inválido',
+    invalidToken: 'El código ingresado no es válido',
```
Comment (mwolman) [inline src/locales/messages/es-ES.js:23]:
```suggestion
    invalidToken: 'Código no válido',
```

File: src/locales/messages/es-ES.js [hunk 6]
```diff
@@ -3,34 +3,45 @@ export default {
...
+    next: 'Siguiente',
     pageNotFoundLegend: 'Parece que la página que estás buscando no existe',
     pageNotFoundTitle: '404',
     password: 'Contraseña',
+    passwordsDontMatch: 'Las contraseñas no coinciden',
     reportError: 'Reportar error',
     required: 'Obligatorio',
     resetPassword: 'Recuperar Contraseña',
-    resetPasswordSuccess: 'Se ha enviado un correo con tu nueva contraseña',
+    resetPasswordSuccess: 'Se ha actualizado la contraseña correctamente',
```
Comment (mwolman) [inline src/locales/messages/es-ES.js:36]:
```suggestion
    resetPasswordSuccess: 'Tu contraseña se ha actualizado con éxito',
```

File: src/components/ForgotPassword/ForgotPassword.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  const [resetStep, setResetStep] = useState(RESET_PASSWORD_STEPS.initial);
+  const [token, setToken] = useState();
+
+  const handleStepChange = (step) => setResetStep(step);
+
+  const handleSaveToken = (newToken) => setToken(newToken);
+
+  return (
+    <AuthWrapper title={intl.messages['common.resetPassword']}>
+      <>
```
Comment (mconiglio) [inline src/components/ForgotPassword/ForgotPassword.js:26]: this fragment is not necessary, right?

File: src/components/App/App.module.scss [hunk 8]
```diff
@@ -1,3 +1,9 @@
+@import '../../tailwind';
+
 body {
   font-family: 'Open Sans', sans-serif;
 }
+
+.loading {
+  @extend .h-screen;
+}
```
Comment (mwolman) [inline src/components/App/App.module.scss:9]: Add EOF please

File: src/components/Form/Input.js [hunk 9]
```diff
@@ -29,6 +29,7 @@ const FormInput = ({
         />
         <input
           {...rest}
+          autoComplete="on"
```
Comment (mconiglio) [inline src/components/Form/Input.js:32]: why? maybe you don't want autocomplete on some inputs, and how it currently is there is no way to override it

File: src/components/ForgotPassword/ForgotPassword.js (hunk 7 above)
Comment (emmaielle) [reply src/components/ForgotPassword/ForgotPassword.js:26]: yeah. Props validation expects only one element, not an array. If I don't add the fragment it thinks it's an array even when they are conditionally rendered and only one renders at a time

File: src/components/Form/Input.js (hunk 9 above)
Comment (emmaielle) [reply src/components/Form/Input.js:32]: I got hyped fixing console warnings. I'll move it back how it was then
